ext/msgpack/unpacker.c

```diff
@@ -168,12 +170,31 @@ static inline int object_complete_ext(msgpack_unpacker_t* uk, int ext_type, VALU
 
     VALUE proc = msgpack_unpacker_ext_registry_lookup(uk->ext_registry, ext_type);
     if(proc != Qnil) {
-        VALUE obj = rb_funcall(proc, s_call, 1, str);
-        return object_complete(uk, obj);
+        long arity = 0;
+        if (rb_obj_class(proc) == rb_cProc) {
+            arity = rb_proc_arity(proc);
+        } else if (rb_obj_class(proc) == rb_cMethod) {
+            arity = FIX2LONG(rb_funcall(proc, s_arity, 0));
+        }
+
+        if (arity == 2) {
+            uk->last_object = Qnil;
+            reset_head_byte(uk);
+            VALUE obj = rb_funcall(proc, s_call, 2, str, uk->buffer.owner);
+
+            if(uk->freeze) {
+                rb_obj_freeze(obj);
+            }
+            uk->last_object = obj;
+            return PRIMITIVE_OBJECT_COMPLETE;
```

Comment:
Why don't you call `object_complete(uk, obj)` here instead of executing a part of it?
I'm not sure about this, but It seems better to do `reset_head_byte(uk);` after `rb_funcall(...)` because the state of `uk` should be clean at the end of deserialization under the unpacker Proc's responsibility.

Reply:
IIRC it's because I need the `reset_head_byte` to be called, otherwise any write the proc may do is broken. I'll double check though as it's been a few days already.

Reply:
Yeah I confirm, here's how it fail if I call the proc immediately:

```
  1) MessagePack::Factory#register_type registering ext type with manual serialization can receive the packer as argument (proc)
     Failure/Error: attrs = unpacker.read
     
     EOFError:
       end of buffer reached
     # ./spec/factory_spec.rb:340:in `read'
     # ./spec/factory_spec.rb:340:in `block (5 levels) in <top (required)>'
     # ./lib/msgpack/factory.rb:70:in `full_unpack'
     # ./lib/msgpack/factory.rb:70:in `load'
     # ./spec/factory_spec.rb:347:in `block (4 levels) in <top (required)>'
```

Reply:
I can understand it should call `reset_head_byte(uk);` before calling the proc.
And I also think that it looks better to call it again after returning from calling the proc. And if it's not required, calling `reset_head_byte(uk);` twice shouldn't be so serious overhead (but just calling `object_complete()` looks very clean).

Reply:
You are right, it's much cleaner this way.